assemble/Particle_Diagnostics.F90

```diff
@@ -0,0 +1,1558 @@
+          temp_part => particle_lists(group_arrays(i))%last
+          if (associated(temp_part)) then
+             allocate(weighted_attributes(i)%col(size(temp_part%attributes)))
+             allocate(weighted_old_attributes(i)%col(size(temp_part%old_attributes)))
+             allocate(weighted_old_fields(i)%col(size(temp_part%old_fields)))
+             weighted_attributes(i)%col(:)=0
+             weighted_old_attributes(i)%col(:)=0
+             weighted_old_fields(i)%col(:)=0
+          end if
+       end do
+    end if
+    node_loc = node_val(xfield,node_num)
+    ele_val(:,:,:) = 0
+    part_total = 0
+    !Loop over elements adjacent to control volume
+    do j = 1,size(ele_nums)
+       do k = 1,xfield%dim+1 !loop over each node of the element
+          if (node_numbers(j,k)==node_num) cycle !cycle if node is the node from our control volume
+          do i = 1,size(group_arrays) !loop over particle grouos
+             temp_part => node_particles(i,node_numbers(j,k))%first
+             allocate(distance(node_particles(i,node_numbers(j,k))%length))
+             distance(:)=0
+             !loop over particles in this group and determine distance from the CV node
+             do l = 1,node_particles(i,node_numbers(j,k))%length
+                do m = 1,xfield%dim
+                   distance(l) = distance(l) + abs(node_loc(m)-temp_part%position(m))**2
+                end do
+                distance(l) = SQRT(distance(l))
+                ele_val(i,j,k) = ele_val(i,j,k) + 1/distance(l)**2!store distance values for weighting
+                if (copy_parents.eqv..false.) then !copy attributes for weighting
+                   weighted_attributes(i)%col(:) = weighted_attributes(i)%col(:) + temp_part%attributes(:)
+                   weighted_old_attributes(i)%col(:) = weighted_old_attributes(i)%col(:) + temp_part%old_attributes(:)
+                   weighted_old_fields(i)%col(:) = weighted_old_fields(i)%col(:) + temp_part%old_fields(:)
+                end if
+                temp_part => temp_part%temp_next
+             end do
+             part_total = part_total + node_particles(i,node_numbers(j,k))%length
+             if (copy_parents.eqv..false.) then !determine the number of particles in each group being weighted
+                length_group(i) = length_group(i) + node_particles(i,node_numbers(j,k))%length
+             end if
+             deallocate(distance)
+          end do
+       end do
+    end do
+
+    !Return if no particles in surrounding control volumes
+    if (part_total==0) then
+       ewrite(2,*) "There are no particles present in adjacent CV's"
+       return
+    end if
+
+    !If adjacent CV's contain particles, clone weighted particles
+    !from adjacent CV's into this CV
+    do i = 1,size(group_arrays)!Loop over particle groups
+       temp_part => particle_lists(group_arrays(i))%last
+       if (associated(temp_part)) then
+
+          id = particle_lists(group_arrays(i))%proc_part_count
```

Comment:
Not necessarily, particles may exist for one particle subgroup but not another. Since we are looping over all groups here, we may come across empty ones that will not be set.

Reply:
Aha, yep. But again this won't be an issue because if `temp_part` isn't associated, the next loop won't be doing anything useful anyway?

Reply:
Yes that's true. Now we aren't getting attribute sizes from them the if condition is redundant.